### PR TITLE
feat(python, rust): timestampNtz support

### PR DESCRIPTION
**Pull request by ion-elgreco**
# Description

- This addresses all our timestamp inconsistencies, where we were reading Primitive:timestamp as a datatetime without UTC, and now we can properly write datetimes with no timezone as columns to Primitive::timestampNtz.
- addressing small bug where checkConstraints feature was not set in writerFeatures when you are on table writer version 7.
- bumping default protocol to 3,7
- Made the pyarrow writer and reader more flexible so we can write/read a 3,7 table as long as it has the supported features there.
- Properly parses timestamps with UTC into pyarrow timestamps with UTC
- Added configkey translation to tablefeature inside the Create Operation

# Related Issue(s)
- closes https://github.com/delta-io/delta-rs/issues/1598
- closes https://github.com/delta-io/delta-rs/issues/1019
- closes https://github.com/delta-io/delta-rs/issues/1777

### CODE REVIEW

File: crates/core/src/operations/transaction/protocol.rs [hunk 1]
```diff
@@ -61,11 +61,11 @@ impl ProtocolChecker {
     }
 
     pub fn default_reader_version(&self) -> i32 {
-        1
+        3
     }
 
     pub fn default_writer_version(&self) -> i32 {
-        2
+        7
```
Comment (rtyler) [inline crates/core/src/operations/transaction/protocol.rs:68]: This is a non-trivial change in behavior isn't it? I don't think timezone support gets us 3/7

Comment (ion-elgreco) [reply crates/core/src/operations/transaction/protocol.rs:68]: Correct, but it smoothes out our table creation versioning since feature toggling in 3,7 is more flexible

Comment (ion-elgreco) [reply crates/core/src/operations/transaction/protocol.rs:68]: Hmm, what's your worry? 

I don't see the issue of automatically enabling a feature during create based on the source data. It's something we can easily infer, if it contains a timezone, then enable the feature and bump the protocol versions.

What I can do, is undo the change of the default protocol version back to 1,2. However I still believe it's more convenient if it's by default on 3,7 especially since we started supporting 3,7 so that we could pick and choose on which table features we added support instead of going linear from 2 to 7.

File: crates/core/src/kernel/expressions/scalars.rs [hunk 2]
```diff
@@ -433,6 +445,16 @@ impl PrimitiveType {
                     .ok_or(self.parse_error(raw))?;
                 Ok(Scalar::Timestamp(micros))
             }
+            TimestampNtz => {
+                let timestamp = NaiveDateTime::parse_from_str(raw, "%Y-%m-%d %H:%M:%S%.f")
```
Comment (hntd187) [inline crates/core/src/kernel/expressions/scalars.rs:449]: So the rough part of this is that the format for the timestamp can be either `1970-01-01 00:00:00`, or `1970-01-01 00:00:00.123456`, I don't believe this format would cover both right?

File: crates/core/src/kernel/models/schema.rs [hunk 3]
```diff
@@ -554,6 +556,7 @@ impl Display for PrimitiveType {
             PrimitiveType::Binary => write!(f, "binary"),
             PrimitiveType::Date => write!(f, "date"),
             PrimitiveType::Timestamp => write!(f, "timestamp"),
+            PrimitiveType::TimestampNtz => write!(f, "timestampNTZ"),
```
Comment (hntd187) [inline crates/core/src/kernel/models/schema.rs:559]: nit: NTZ -> Ntz

File: crates/core/src/operations/create.rs [hunk 4]
```diff
@@ -233,8 +237,36 @@ impl CreateBuilder {
...
+            .filter(|v| match v {
+                ReaderFeatures::Other(_) => false,
+                _ => true,
+            })
+            .collect::<HashSet<ReaderFeatures>>();
+
+        let contains_timestampntz = &self
+            .columns
+            .iter()
+            .any(|f| f.data_type() == &DataType::TIMESTAMPNTZ);
```
Comment (hntd187) [inline crates/core/src/operations/create.rs:263]: Should the enum variant be all caps here?

File: crates/core/src/kernel/expressions/scalars.rs (hunk 2 above)
Comment (ion-elgreco) [reply crates/core/src/kernel/expressions/scalars.rs:449]: Not sure, I saw the mention in the protocol on both formats. But timestamp was only parsed with the ms precision. Maybe @roeap can chip in here?

File: crates/core/src/kernel/models/schema.rs (hunk 3 above)
Comment (ion-elgreco) [reply crates/core/src/kernel/models/schema.rs:559]: I am not sure anymore for what exactly, but I had to do this on purpose

**Comment by hntd187** [comment]
Also delta-scala doesn't universally cast things to UTC, it respects the `isAdjustedToUTC` param for timestamps, where timestampntz has this set to false, while normal timestamps are set to true

File: crates/core/src/kernel/expressions/scalars.rs (hunk 2 above)
Comment (hntd187) [reply crates/core/src/kernel/expressions/scalars.rs:449]: https://github.com/delta-io/delta/blob/master/PROTOCOL.md#timestamp-without-timezone-timestampNtz it talks about supporting both formats for this datatype here

File: crates/core/src/operations/create.rs (hunk 4 above)
Comment (ion-elgreco) [reply crates/core/src/operations/create.rs:263]: This points to the pub const on DataType:

`pub const TIMESTAMPNTZ: Self = DataType::Primitive(PrimitiveType::TimestampNtz);`

Comment (hntd187) [reply crates/core/src/operations/create.rs:263]: missed this, never mind :-D

File: crates/core/src/kernel/expressions/scalars.rs (hunk 2 above)
Comment (ion-elgreco) [reply crates/core/src/kernel/expressions/scalars.rs:449]: I checked, it has no problem parsing both formats with "%Y-%m-%d %H:%M:%S%.f", both can be parsed into a timestamp

File: crates/core/src/operations/transaction/protocol.rs (hunk 1 above)
Comment (hntd187) [reply crates/core/src/operations/transaction/protocol.rs:68]: because doing this breaks all readers/writers in the wild that aren't 3/7, it breaks prior versions of delta-spark as well, even if you never wrote a single row to the table because it requires readers and writers to meet that bar to even read the table. I'm saying, we should only update the protocol if the features are explicitly enabled or explicitly used. For example, databricks enables deletion vectors by default which makes basically the default protocol 3,7, but if you disable that in the table create it sets the protocol to 1,2. The protocol versions should be the absolute minimum required to read/write the table, we should never just update them if there isn't a feature that requires it.

File: crates/core/src/kernel/models/schema.rs (hunk 3 above)
Comment (ion-elgreco) [reply crates/core/src/kernel/models/schema.rs]: Was not needed after all xD

File: crates/core/src/operations/transaction/protocol.rs (hunk 1 above)
Comment (ion-elgreco) [reply crates/core/src/operations/transaction/protocol.rs:68]: Gotcha, I see your point now, that will definitely break backwards compatibility indeed with older databricks runtimes. 

Then I'll change back the default to 1,2 but bump it only  during a protocol commit when we see a feature that is only on 3,7.

Comment (hntd187) [reply crates/core/src/operations/transaction/protocol.rs:68]: Excellent, so when you create a table with timestamp_ntz, it will create a table with 3,7 as the versions and timestampNtz in the table_features

Comment (ion-elgreco) [reply crates/core/src/operations/transaction/protocol.rs]: I have undone the changes :) So it's now only done while writing a table with no timezone.

File: crates/core/src/kernel/models/actions.rs [hunk 5]
```diff
@@ -281,15 +283,15 @@ impl From<String> for WriterFeatures {
...
     fn from(value: &str) -> Self {
         match value {
-            "appendOnly" => WriterFeatures::AppendOnly,
-            "invariants" => WriterFeatures::Invariants,
-            "checkConstraints" => WriterFeatures::CheckConstraints,
-            "changeDataFeed" => WriterFeatures::ChangeDataFeed,
+            "appendOnly" | "delta.appendOnly" => WriterFeatures::AppendOnly,
+            "invariants" | "delta.invariants" => WriterFeatures::Invariants,
+            "checkConstraints" | "delta.checkConstraints" => WriterFeatures::CheckConstraints,
+            "changeDataFeed" | "delta.enableChangeDataFeed" => WriterFeatures::ChangeDataFeed,
```
Comment (rtyler) [inline crates/core/src/kernel/models/actions.rs:289]: @ion-elgreco why are these changes required? Are we not picking up table properties correctly right now? If so, is there a bug this is related to?

Comment (ion-elgreco) [reply crates/core/src/kernel/models/actions.rs:289]: @rtyler I am doing that so I can go from tableConfigKeys into tableFeatures. This is what spark-delta also does if your table protocol versions are 3,7 and you do set tblproperties "delta.enableChangeDataFeed" = true, it will additionally set the feature ChangeDataFeed into the protocol commit.

File: crates/core/src/kernel/expressions/scalars.rs [hunk 6]
```diff
@@ -222,10 +225,17 @@ impl Scalar {
                 .downcast_ref::<Date32Array>()
                 .map(|v| Self::Date(v.value(index))),
             // TODO handle timezones when implementing timestamp ntz feature.
-            Timestamp(TimeUnit::Microsecond, None) => arr
-                .as_any()
-                .downcast_ref::<TimestampMicrosecondArray>()
-                .map(|v| Self::Timestamp(v.value(index))),
+            Timestamp(TimeUnit::Microsecond, tz) => match tz {
```
Comment (guilhem-dvr) [inline crates/core/src/kernel/expressions/scalars.rs:228]: nit: the CL addresses the `TODO` no?

Comment (ion-elgreco) [reply crates/core/src/kernel/expressions/scalars.rs:228]: Right, forgot to remove that : )

**Review by rtyler — APPROVED** [review]

File: crates/core/src/kernel/models/actions.rs (hunk 5 above)
Comment (roeap) [reply crates/core/src/kernel/models/actions.rs:289]: hmm, just passing by, but this seems strange. these are the values written in the feature vector list, so this seems like a strange place to do this conversion as well, isnce it describes completely differnet fields, now? one is values in the feature vecotor field, andother one is keys in the table properties, which should be handled in the table config?

File: crates/core/src/kernel/models/schema.rs [hunk 7]
```diff
@@ -468,7 +468,7 @@ fn default_true() -> bool {
 }
 
 #[derive(Debug, Serialize, Deserialize, PartialEq, Clone, Eq, Hash)]
-#[serde(rename_all = "camelCase")]
+#[serde(rename_all = "snake_case")]
```
Comment (roeap) [inline crates/core/src/kernel/models/schema.rs:471]: are we sure this is right?

Comment (ion-elgreco) [reply crates/core/src/kernel/models/schema.rs:471]: Without this it would fail for the timestampNtz, since spark wrote it as timestamp_ntz in the log. I couldnt find his anywhere in the docs, but it only impacted timestampNtz since the rest of them are one word long

File: crates/core/src/kernel/models/actions.rs (hunk 5 above)
Comment (ion-elgreco) [reply crates/core/src/kernel/models/actions.rs:289]: I added the table config keys here as well for quick parsing to table features, see my comment above.

In spark when your protocol is at 3,7 it will auto convert the configs into table features so that's basically what I am doing here as well.
